pytest_bdd/cucumber_json.py

```diff
@@ -106,15 +148,15 @@ def pytest_runtest_logreport(self, report):
             # skip if there isn't a result or scenario has no steps
             return
 
-        def stepmap(step):
+        def stepmap(step, step_names):
             error_message = False
             if step['failed'] and not scenario.setdefault('failed', False):
                 scenario['failed'] = True
                 error_message = True
 
             return {
                 "keyword": step['keyword'],
-                "name": step['name'],
+                "name": next(step_names),
```

Comment:
Why don't you format it right here? This function is addressing "scenario" from the parent scope, you could just call format_name from here.
Please also introduce a flag --cucumber-json-expanded for it.

I see now that all our pytest items have their own scenario report with only one row of examples. Well.. We have to think how to combine them back in the future.

So let's not break the compatibility by this change and lets replace <params> only if the --cucumber-json-expanded is on. We'll think about fixing the original report format later.

Reply:
Formatting:
The problem is getting the right variables at the right time to format the step names. Maybe I should document that better in `StepNames`.

Basically: when you have an outline scenario with 2 examples, `LogBDDCucumberJSON.pytest_runtest_logreport()` is called twice. From within `step_names()` method you have no idea which of the 2 is run, so which set of variables do you choose? You have access to the scenario (and thus the examples), but no idea which index to use.

By saving the state in the generators, we ensure that we get the fixture names in the right order.

Flag:
Are you sure we need this `--cucumber-json-expanded` flag? I think the current output is already "expanded" since it doesn't include the examples and instead has all the steps "expanded". The only thing that is not done according to the specs is the filling in of the variable names.

I'll take a look to add this, but I'm not entirely sure that this would be the right name, since in my opinion the format is 'expanded' by default.

Reply:
I think we need to fix the "original" format instead. You can always pass expanded on your CI, right?
I still think StepNames is a hack and we need to refactor the internal structure instead. For now you can rely on fact that there's only one row of examples for this particular scenario. Later we could lookup the row of examples and use it while formatting steps.

Reply:
Thanks for your response :)

How do you mean "fix the original format"? I can interpret this in two ways:

1) You mean change the cucumber output so it uses the "non-expand" output -> This will totally change the output, not using "scenario" anymore, but outputting "scenario_outlines" and "examples", plus you'll lose the results, this is not what you want I think, so I'm thinking you mean option 2 instead:
2) You mean update the cucumber output so it used the "expand" output -> this is what I've done in this PR.

About the StepNames, I agree that this is not the nicest implementation because it hides some complexity (ie. the generators generate names in parallel with the `pytest_runtest_logreport()` and it only works because of some assumptions: the most important one "the steps are reported in a particular order".

Not sure what you mean with "rely on fact that there's only one row of examples for this particular scenario"? You can see that 3 rows are added in this scenario-outline example: https://github.com/pytest-dev/pytest-bdd/pull/226/files#diff-4c2da9247616053392e9474b89319cb3R56